sdk/src/azure/core/az_mqtt5_rpc.c

```diff
@@ -21,6 +23,249 @@ static const az_span command_name_key = AZ_SPAN_LITERAL_FROM_STR(AZ_MQTT5_RPC_CO
+static const uint32_t az_mqtt5_rpc_service_id_hash = 4175641829;
+static const uint32_t az_mqtt5_rpc_executor_id_hash = 3913329219;
+static const uint32_t az_mqtt5_rpc_command_id_hash = 2624200456;
+
+AZ_INLINE uint32_t _az_mqtt5_rpc_calculate_hash(az_span token)
+{
+  uint32_t hash = 5831;
+  for (int32_t i = 0; i < az_span_size(token); i++)
+  {
+    hash = ((hash << 5) + hash) + az_span_ptr(token)[i];
+  }
+  return hash;
+}
+
+AZ_INLINE az_span _az_mqtt5_rpc_extract_next_token_from_format(az_span format)
+{
+  int32_t pos = az_span_find(format, token_format_end_character);
+  if (pos == -1)
+  {
+    return AZ_SPAN_EMPTY;
+  }
+  return az_span_slice(format, 0, pos);
+}
+
+AZ_INLINE az_span _az_mqtt5_rpc_extract_next_token_from_topic(az_span format)
+{
+  int32_t pos = az_span_find(format, token_topic_end_character);
+  if (pos == -1)
+  {
+    return az_span_slice_to_end(format, 0);
+  }
+  return az_span_slice(format, 0, pos);
+}
+
+AZ_INLINE az_result _az_mqtt5_rpc_verify_buffer_length_and_copy(
+    az_span value,
+    az_span* remainder,
+    uint32_t* required_length,
+    uint32_t buffer_length)
+{
+  if (!_az_span_is_valid(value, 1, false))
+  {
+    return AZ_ERROR_ARG;
+  }
+  *required_length += (uint32_t)az_span_size(value) - 1;
+  if (*required_length > (uint32_t)buffer_length)
+  {
+    return AZ_ERROR_NOT_ENOUGH_SPACE;
+  }
+  *remainder = az_span_copy(*remainder, value);
+
+  return AZ_OK;
+}
+
+AZ_INLINE az_result
+_az_mqtt5_rpc_verify_match_and_copy(az_span value, az_span extracted_token, az_span* out_token)
+{
+  if (az_span_is_content_equal(value, AZ_SPAN_EMPTY)
+      || az_span_is_content_equal(value, extracted_token))
+  {
+    if (out_token)
+    {
+      *out_token = extracted_token;
+    }
+    return AZ_OK;
+  }
+  return AZ_ERROR_IOT_TOPIC_NO_MATCH;
+}
+
+AZ_NODISCARD az_result _az_mqtt5_rpc_replace_tokens_in_format(
+    az_span mqtt_topic_span,
+    az_span topic_format,
+    az_span service_group_id,
+    az_span client_id,
+    az_span service_id,
+    az_span executor_id,
+    az_span command_id,
+    uint32_t* required_length)
+{
+  az_result ret = AZ_OK;
+  *required_length += (int32_t)sizeof((uint8_t)'\0');
+  uint32_t buffer_length = (uint32_t)az_span_size(mqtt_topic_span);
+  az_span remainder = mqtt_topic_span;
+
+  if (!az_span_is_content_equal(service_group_id, AZ_SPAN_EMPTY))
+  {
+    *required_length += (uint32_t)az_span_size(service_group_id_key)
+        + (uint32_t)az_span_size(service_group_id) + 1;
+    _az_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_span, (int32_t)*required_length);
+    remainder = az_span_copy(remainder, service_group_id_key);
+    remainder = az_span_copy(remainder, service_group_id);
+    remainder = az_span_copy_u8(remainder, '/');
+  }
+
+  for (int32_t i = 0; i < az_span_size(topic_format); i++)
+  {
+    uint8_t c = az_span_ptr(topic_format)[i];
+    _az_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_span, (int32_t)(++*required_length));
```

Comment:
(Adding this just for discussion) I know it isn't possible with this solution, but I think it would be better to calculate the required length before copying into the out memory instead of the possibility of partially writing the topic and then failing and returning.

Reply:
Adding @CIPop to the discussion, if we were to calculate the required length prior to doing any copying we'd have to do two passes at which point we wouldn't need the hashing function. On the first pass I can identify where all the tokens are and on the second replace them.

Reply:
As long as we guarantee safety, it's OK to attempt to fill part of the buffer.